llvm/lib/Target/RISCV/MCA/RISCVCustomBehaviour.cpp

```diff
@@ -14,12 +14,33 @@
 #include "RISCVCustomBehaviour.h"
 #include "MCTargetDesc/RISCVMCTargetDesc.h"
 #include "RISCV.h"
+#include "RISCVISelDAGToDAG.h"
 #include "TargetInfo/RISCVTargetInfo.h"
 #include "llvm/MC/TargetRegistry.h"
 #include "llvm/Support/Debug.h"
 
 #define DEBUG_TYPE "llvm-mca-riscv-custombehaviour"
 
+namespace llvm::RISCV::mca {
+struct VXMemOpInfo {
+  unsigned Log2IdxEEW : 3;
+  unsigned IsOrdered : 1;
+  unsigned IsStore : 1;
+  unsigned NF : 4;
+  unsigned BaseInstr;
+};
+
+#define GET_RISCVBaseVXMemOpTable_DECL
+#define GET_RISCVBaseVXMemOpTable_IMPL
+// We need to include the implementation code here because RISCVCustomBehavior
+// is not linked against RISCVCodeGen.
+#define GET_RISCVVLXSEGTable_IMPL
```

Comment:
So the declarations are in `llvm::RISCV` namespace but the implementations are in `llvm::RISCV::mca`? How does that even work?

Reply:
I guess the DECLs of the function in RISCVISelDAGToDAG.h are being declared in the other namespace but the implementation is in this namespace and the callers are looking in this namespace. The function declarations in RISCVISelDAGToDAG.h are just useless.

And the struct definition work because the llvm::RISCV is a parent of llvm::RISCV::mca

Reply:
Yeah, RISCVISelDAGToDAG.h is included for those structs. The function declarations there are not really relevant.